examples/hubert/dataset/hubert_dataset.py

```diff
@@ -78,56 +89,59 @@ def _get_buckets(
+        boundaries = torch.linspace(min_len, max_len, interval)
+        bucket_ids = torch.bucketize(torch.tensor(lengths), boundaries)
+        for i in range(bucket_ids.size(0)):
             bucket_id = bucket_ids[i]
             if bucket_id in buckets:
                 buckets[bucket_id].append(i)
             else:
                 buckets[bucket_id] = [i]
         for k in buckets:
-            if self.shuffle:
-                random.shuffle(buckets[k])
             buckets[k] = torch.as_tensor(buckets[k], dtype=torch.int)
         return buckets
 
     def __iter__(self) -> Iterator[List[int]]:
+        if self.shuffle:
+            for k in self.buckets:
+                self.buckets[k] = self.buckets[k][torch.randperm(self.buckets[k].size(0))]
+
         iter_list = []
         total_len = 0
         batch = []
-        if self.max_token_count:
-            for k in self.buckets.keys():
-                for i in range(self.buckets[k].size(0)):
-                    index = self.buckets[k][i]
-                    if total_len > self.max_token_count:
-                        iter_list.append(batch)
-                        batch = [index]
-                        total_len = self.lengths[index]
-                    else:
-                        batch.append(index)
+        for k in sorted(self.buckets):
+            for i in range(self.buckets[k].size(0)):
+                index = int(self.buckets[k][i])
+                if self.max_token_count is not None:
+                    if total_len + self.lengths[index] <= self.max_token_count:
+                        batch.append(self.indices[index])
                         total_len += self.lengths[index]
-        else:
-            for k in self.buckets.keys():
-                for i in range(self.buckets[k].size(0)):
-                    index = self.buckets[k][i]
+                    else:
+                        iter_list.append(batch)
+                        batch = []
+                        total_len = 0
+                else:
                     if total_len == self.batch_size:
                         iter_list.append(batch)
-                        batch = [index]
+                        batch = [self.indices[index]]
                         total_len = 1
                     else:
-                        batch.append(index)
+                        batch.append(self.indices[index])
                         total_len += 1
+            if len(batch) > 0 and (self.drop_last or self.max_token_count):
+                iter_list.append(batch)
 
-        for batch in iter_list:
-            yield batch
+        return iter(iter_list)
 
     def __len__(self):
-        return len(self.data_source)
+        if self.batch_size:
+            if self.drop_last:
+                return len(self.lengths) // self.batch_size
+            else:
+                return (len(self.lengths) + self.batch_size - 1) // self.batch_size
+        else:
+            return None
```

Comment:
is this a valid return value, i.e. dataloaders know how to handle it?

Reply:
In the new commit, the `self.iter_list` is initialized before `__iter__` to make sure the length is valid. Then every time the `iter_list` is updated, so will be the value returned by `__len__` method.